src/MichiPointsMinter.sol

```diff
@@ -0,0 +1,170 @@
+
+    /// @notice precision denominator for fees
+    uint256 public precision;
+
+    /// @notice address that receives tokenize fees
+    address public feeReceiver;
+
+    /// @notice mapping of approved tokenized points token addresses
+    mapping(address => bool) private approvedTokenizedPoints;
+
+    /// @notice mapping of fulfilled requests by request id and chain initiated on
+    mapping(uint256 => mapping(uint256 => FulfilledRequest)) public chainToRequestId;
+
+    /// @notice mapping of fulfilled requests by chain
+    mapping(uint256 => uint256[]) public fulfilledRequestsByChain;
+
+    /// @notice mapping of total amount of tokenized points minted to users by tokenized points token address
+    mapping(address => uint256) public userAmountMintedByTokenizedPoint;
+
+    /// @notice mapping of total amount of fees taken by tokenized points token address
+    mapping(address => uint256) public feesByTokenizedPoint;
+
+    /// @notice error when array length is not matching
+    error ArrayLengthMismatch();
+
+    /// @notice error when unapproved token is inputted as parameter
+    error UnapprovedToken(address token);
+
+    /// @notice error when 0 tokens are requested to be minted
+    error InvalidAmount(uint256 amount);
+
+    /// @notice error when request can already been fulfilled
+    error RequestAlreadyFulfilled(uint256 chainId, uint256 requestId);
+
+    /// @notice error when tokenize fee is set too high
+    error InvalidTokenizeFee(uint256 tokenizeFee);
+
+    /// @notice error when approving an already approved token
+    error TokenizedPointAlreadyApproved(address tokenizedPoint);
+
+    /// @notice error when removing an unapproved token
+    error TokenizedPointNotApproved(address tokenizedPoint);
+
+    /// @notice event when a tokenize point request has been completed
+    event RequestFulfilled(
+        address receier, address[] tokenizedPoints, uint256 chainId, uint256 requestId, uint256[] amounts
+    );
+
+    /// @dev Constructor for MichiPointsMinter contract
+    /// @param feeReceiver_ address of the fee receiver
+    /// @param tokenizeFee_ tokenize fee
+    /// @param precision_ precision denominator
+    constructor(address feeReceiver_, uint256 tokenizeFee_, uint256 precision_) {
+        _setupRole(DEFAULT_ADMIN_ROLE, msg.sender);
+        _setupRole(MINTER_ROLE, msg.sender);
+
+        feeReceiver = feeReceiver_;
+        tokenizeFee = tokenizeFee_;
+        precision = precision_;
+    }
+
+    /// @dev Initiate tokenized points minting
+    /// @param receiver address receiving the minted tokens
+    /// @param tokensToMint array of tokens to mint
+    /// @param amounts array of amounts to mint
+    /// @param chainId chain that request was made on
+    /// @param requestId id of request
+    function mintTokenizedPoints(
+        address receiver,
+        address[] calldata tokensToMint,
+        uint256[] calldata amounts,
+        uint256 chainId,
+        uint256 requestId
```

Comment:
does it make sense to also include address of the MichiTokenizeRequestor? This way if you ever need to upgrade you can make sure a request is not duplicated

Reply:
if we use an upgradable contract design, wouldn't the contract address stay the same and _requestId would continue and not reset back to 0?